services/src/datasets/external/sentinel_s2_l2a_cogs.rs

```diff
@@ -0,0 +1,640 @@
+use crate::stac::Feature as StacFeature;
+use crate::stac::FeatureCollection as StacCollection;
+use crate::stac::StacAsset;
+use crate::{datasets::listing::DatasetListOptions, error::Result};
+use crate::{
+    datasets::{
+        listing::{DatasetListing, DatasetProvider},
+        storage::DatasetProviderDefinition,
+    },
+    error,
+    util::user_input::Validated,
+};
+use async_trait::async_trait;
+use chrono::{DateTime, Duration, Utc};
+use geoengine_datatypes::dataset::{DatasetId, DatasetProviderId, ExternalDatasetId};
+use geoengine_datatypes::primitives::{Measurement, TimeInterval};
+use geoengine_datatypes::raster::GeoTransform;
+use geoengine_datatypes::raster::RasterDataType;
+use geoengine_datatypes::spatial_reference::{SpatialReference, SpatialReferenceAuthority};
+use geoengine_operators::engine::QueryRectangle;
+use geoengine_operators::source::GdalDatasetParameters;
+use geoengine_operators::source::GdalLoadingInfoPart;
+use geoengine_operators::source::GdalLoadingInfoPartIterator;
+use geoengine_operators::{
+    engine::{MetaData, MetaDataProvider, RasterResultDescriptor, VectorResultDescriptor},
+    mock::MockDatasetDataSourceLoadingInfo,
+    source::{GdalLoadingInfo, OgrSourceDataset},
+};
+use log::debug;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::collections::HashMap;
+use std::path::PathBuf;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SentinelS2L2ACogsProviderDefinition {
+    name: String,
+    id: DatasetProviderId,
+    api_url: String,
+}
+
+#[typetag::serde]
+#[async_trait]
+impl DatasetProviderDefinition for SentinelS2L2ACogsProviderDefinition {
+    async fn initialize(
+        self: Box<Self>,
+    ) -> crate::error::Result<Box<dyn crate::datasets::listing::DatasetProvider>> {
+        Ok(Box::new(SentinelS2L2aCogsDataProvider::new(
+            self.id,
+            self.api_url,
+        )))
+    }
+
+    fn type_name(&self) -> String {
```

Comment:
whats this? / why is this needed? why is it not static?

Reply:
it is the type of the provider, while the name ist the concrete instance. "WCS" vs. "DWD Geoserver".
It cannot be static because then it wouldn't be deserializable with type_tag